rs/xnet/payload_builder/src/lib.rs

```diff
@@ -879,11 +883,24 @@ impl XNetPayloadBuilderImpl {
 ///
 /// In order to prevent mutual stalling, only applies to incoming NNS
 /// streams; and to `Application`-subnet-to-`System`-subnet streams.
-pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usize> {
+pub fn get_msg_limit(
```

Comment:
The doc comment needs updating.

On a second thought, I would rather have a separate function (or just inline code) to compute an upper bound for `slice.end` (which is just `stream.begin + 50k`, nothing fancy). As is, I'm not 100% sure about this, but we may not even enforce the 50k limit for application subnet streams on system subnets.

It would be a lot easier if we kept this as is; and then, where we call it, simply took the minimum of whatever this returns and `stream_begin + 50k - expected.message_index`. Then you're obviously under both limits at all times.

Reply:
Made it its own function `get_signal_limit()` that follows the logic you suggest here.